src/dune_rules/pkg_rules.ml

```diff
@@ -352,31 +351,50 @@ end
 
 module Substitute = struct
   module Spec = struct
-    type ('path, 'target) t = 'path * 'target
+    type ('path, 'target) t =
+      string Substs.Map.t * Dune_lang.Package_name.t * 'path * 'target
```

Comment:
This is one way to do it, but it's actually less efficient. In particular, we're doing all the work to produce this map when generating the action, but we might not necessarily run the action. Instead of generating this map, we can produce the expander when running the action in the same as we do it elsewhere in the code - by running `Action_expander.expander`.

So all we need to save into the action are the inputs required to call this function.

Or, we can even save the expander that we used at rule generation and fetch it from there.

Reply:
I thought about this as well. In particular `misc` is not available in the substitutions as calling it would raise an error and wrapping everything in lazy doesn't help as encoding would trigger the `Code_error` in `Paths.get` anyway when encoding.

The problem with saving the expander is that `encode`ing the expander is encoding a lot of irrelevant stuff, the issue with creating an expander is that even if I save the `Context_name.t` and `Pkg.t` to create an expander using `expander`, it will miss out on the `Withenv`'d values as these get added to the expander as it traverses the syntax tree.